source/sixtrack.f90

```diff
@@ -1680,473 +1671,375 @@ subroutine wzsub(x,y,u,v)
+            aaiv(k,i) = scalemu(im)*(ak0(im,k)+amultip(k,i)*aka(im,k))
+            bbiv(k,i) = scalemu(im)*(bk0(im,k)+bmultip(k,i)*bka(im,k))
+          end do
+        end if
+      end do
+    end if
 
-!--Cavities (ktrack = 2 for thin)
-      elseif(abs(kz(ix)).eq.12) then
-         !Moved from daten
-         phasc(ix) = el(ix)
-         el(ix) = zero
-         dynk_elemData(ix,3) = phasc(ix)
-         if (.not.lfirst) then
-
-            ! Doesn't work, as i is not initialized here.
-            !if (.not.ktrack(i).eq.2) goto 100 !ERROR
-
-            phasc(ix) = phasc(ix)*rad
-
-            hsyc(ix) = ((two*pi)*ek(ix))/tlen         ! daten SYNC block
-            hsyc(ix)=(c1m3*hsyc(ix))*real(itionc(ix),fPrec) ! trauthin/trauthck
-         endif
-!--BEAM-BEAM
-      elseif(kz(ix).eq.20) then
-
-        if (lfirst) then
-          ptnfac(ix)=el(ix)
-          el(ix)=zero
-          parbe(ix,5) = ed(ix)
-          ed(ix)=zero
-          parbe(ix,6) = ek(ix)
-          ek(ix)=zero
-          endif
-! This is to inialize all the beam-beam element before the tracking (or to update it for DYNK).
-        if (.not.lfirst) then
-          do i=1,iu
-            if ( ic(i)-nblo.eq.ix ) then
-              ibb=imbb(i)
-              if(parbe(ix,2).gt.zero) then
-                if(beam_expflag.eq.1) then
-                 bbcu(ibb,1)=parbe(ix,7)
-                 bbcu(ibb,4)=parbe(ix,8)
-                 bbcu(ibb,6)=parbe(ix,9)
-                 bbcu(ibb,2)=parbe(ix,10)
-                 bbcu(ibb,9)=parbe(ix,11)
-                 bbcu(ibb,10)=parbe(ix,12)
-                 bbcu(ibb,3)=parbe(ix,13)
-                 bbcu(ibb,5)=parbe(ix,14)
-                 bbcu(ibb,7)=parbe(ix,15)
-                 bbcu(ibb,8)=parbe(ix,16)
-                  do ii=1,10
-                    bbcu(ibb,ii)=bbcu(ibb,ii)*c1m6
-                  enddo
-                endif
-                ktrack(i)=44
-                parbe(ix,4)=(((-one*crad)*ptnfac(ix))*half)*c1m6
-                if(ibeco.eq.1) then
-                  track6d(1,1)=parbe(ix,5)*c1m3
-                  track6d(2,1)=zero
-                  track6d(3,1)=parbe(ix,6)*c1m3
-                  track6d(4,1)=zero
-                  track6d(5,1)=zero
-                  track6d(6,1)=zero
-                  napx0=napx
-                  napx=1
-                  call beamint(napx,track6d,parbe,sigz,bbcu,imbb(i),ix,ibtyp,ibbc, mtc)
-                  beamoff(1,imbb(i))=track6d(1,1)*c1e3
-                  beamoff(2,imbb(i))=track6d(3,1)*c1e3
-                  beamoff(3,imbb(i))=track6d(5,1)*c1e3
-                  beamoff(4,imbb(i))=track6d(2,1)*c1e3
-                  beamoff(5,imbb(i))=track6d(4,1)*c1e3
-                  beamoff(6,imbb(i))=track6d(6,1)
-                  napx=napx0
+  ! Cavities (ktrack = 2 for thin)
+  elseif(abs(kz(ix)) == 12) then
+    dynk_elemData(ix,3) = el(ix)
+    phasc(ix) = el(ix)*rad
+    el(ix) = zero
+    if(lfirst) then
+      if(abs(ed(ix)) > pieni .and. abs(ek(ix)) > pieni) then
```

Comment:
Hmm. This logic would interfer with DYNK of RF cavities, but I'm not entirely sure how to solve it.

Basically how it is now, I think it will disable the cavity completely (switching to 4D tracking if all cavities are like this) if neither the voltage or frequency is set already here just after parsing the geometry; which is *probably* OK, since if the frequency is unset, then the cavity is really not properly initialized... This should at least be detected in dynk_pretrack or sanitycheck tough, before trying to use the RFCAV.

I believe there is a "if voltage is zero AND not used by DYNK then ktrack=31" at some later point which would make it the main disabling?

There is some history to these kind of fixes; there may be some useful notes in this PR, which fixed a bunch of bugs in the RF cavity parsing: https://github.com/SixTrack/SixTrack/commit/334d0a3e4ae15c9f0403e12e425d1aab47b42bcf

As a side, I wonder if we really need the `kp` array at all? At least for RF cavities... AFAIK they are (where?) also used by aperture.

Reply:
Unless I missed something, this is exactly the same logic as is currently in SixTrack.